src/ORM/Hierarchy/MarkedSet.php

```diff
@@ -0,0 +1,805 @@
+<?php
+
+namespace SilverStripe\ORM\Hierarchy;
+
+use InvalidArgumentException;
+use LogicException;
+use SilverStripe\Core\Injector\Injectable;
+use SilverStripe\ORM\ArrayList;
+use SilverStripe\ORM\DataObject;
+use SilverStripe\ORM\FieldType\DBField;
+use SilverStripe\ORM\SS_List;
+use SilverStripe\View\ArrayData;
+
+/**
+ * Contains a set of hierarchal objects generated from a marking compilation run
+ */
+class MarkedSet
+{
+    use Injectable;
+
+    /**
+     * Marked nodes for a given subtree. The first item in this list
+     * is the root object of the subtree.
+     *
+     * A marked item is an item in a tree which will be included in
+     * a resulting tree.
+     *
+     * @var array Map of [itemID => itemInstance]
+     */
+    protected $markedNodes;
+
+    /**
+     * Optional filter callback for filtering nodes to mark
+     *
+     * Array with keys:
+     *  - parameter
+     *  - value
+     *  - func
+     *
+     * @var array
+     * @temp made public
+     */
+    public $markingFilter;
+
+    /**
+     * @var DataObject
+     */
+    protected $rootNode = null;
+
+    /**
+     * Method to use for getting children. Defaults to 'AllChildrenIncludingDeleted'
+     *
+     * @var string
+     */
+    protected $childrenMethod = null;
+
+    /**
+     * Method to use for counting children. Defaults to `numChildren`
+     *
+     * @var string
+     */
+    protected $numChildrenMethod = null;
+
+    /**
+     * Minimum number of nodes to iterate over before stopping recursion
+     *
+     * @var int
+     */
+    protected $nodeCountThreshold = null;
+
+    /**
+     * Max number of nodes to return from a single children collection
+     *
+     * @var int
+     */
+    protected $maxChildNodes;
+
+    /**
+     * Enable limiting
+     *
+     * @var bool
+     */
+    protected $enableLimiting = true;
+
+    /**
+     * Create an empty set with the given class
+     *
+     * @param DataObject $rootNode Root node for this set. To collect the entire tree,
+     * pass in a singelton object.
+     * @param string $childrenMethod Override children method
+     * @param string $numChildrenMethod Override children counting method
+     * @param int $nodeCountThreshold Minimum threshold for number nodes to mark
+     * @param int $maxChildNodes Maximum threshold for number of child nodes to include
+     */
+    public function __construct(
+        DataObject $rootNode,
+        $childrenMethod = null,
```

Comment:
The problem is that most of the time you have to always add those fluent api calls. I've added the most common to the constructor. This was at @sminnee's request. Originally the constructor only took a single param.